File: my-folder/0039-combination-sum/solution.cpp
class Solution {
public:
    void recursion(vector<int> &candidates,vector<vector<int>> &ans,vector<int>& temp,int sum,int i,int target){
        if(sum==target){
            ans.push_back(temp);
        }
        if(i==candidates.size()|| sum>target) return;
        if(sum<target){
        temp.push_back(candidates[i]);
        recursion(candidates, ans, temp, sum + candidates[i], i, target);
        temp.pop_back();
        recursion(candidates, ans, temp, sum, i + 1, target);

    }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> temp;
        recursion(candidates,ans,temp,0,0,target);
        return ans;
        
    }
};

File: my-folder/0039-combination-sum/solution.java
public class Solution {

    void getList(int[] candidates, int sum, List<List<Integer>> ans, List<Integer> temp, int n) {
        if (sum == 0) {
            ans.add(new ArrayList<>(temp));
            return;
        }
        if (sum < 0 || n <= 0) return;

        temp.add(candidates[n - 1]);
        getList(candidates, sum - candidates[n - 1], ans, temp, n);
        temp.remove(temp.size() - 1);
        getList(candidates, sum, ans, temp, n - 1);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        int n = candidates.length;
        getList(candidates, target, ans, temp, n);
        return ans;
    }
}

